src/dma.rs

```diff
@@ -0,0 +1,699 @@
+    // This is always a `Some` outside of `drop`.
+    inner: Option<TransferInner<B, C, T>>,
+}
+
+impl<B, C: Channel, T> Transfer<B, C, T> {
+    /// Start a DMA write transfer.
+    pub fn start_write(mut buffer: B, mut channel: C, target: T) -> Self
+    where
+        B: WriteBuffer + 'static,
+        T: Target<C>,
+    {
+        let (ptr, len) = buffer.write_buffer();
+
+        channel.set_memory_address(ptr as u32, Increment::Enable);
+        channel.set_transfer_length(len);
+        channel.set_word_size::<B::Word>();
+        channel.set_direction(Direction::FromPeripheral);
+
+        unsafe { Self::start(buffer, channel, target) }
+    }
+
+    /// Start a DMA read transfer.
+    pub fn start_read(buffer: B, mut channel: C, target: T) -> Self
+    where
+        B: ReadBuffer + 'static,
+        T: Target<C>,
+    {
+        let (ptr, len) = buffer.read_buffer();
+
+        channel.set_memory_address(ptr as u32, Increment::Enable);
+        channel.set_transfer_length(len);
+        channel.set_word_size::<B::Word>();
+        channel.set_direction(Direction::FromMemory);
+
+        unsafe { Self::start(buffer, channel, target) }
+    }
+
+    /// # Safety
+    ///
+    /// Callers must ensure that:
+    ///
+    /// - the given buffer will be valid for the duration of the transfer
+    /// - the DMA channel is configured correctly for the given target and buffer
+    unsafe fn start(buffer: B, mut channel: C, target: T) -> Self
+    where
+        T: Target<C>,
+    {
+        assert!(!channel.is_enabled());
+
+        atomic::compiler_fence(Ordering::Release);
+
+        channel.enable();
+
+        Self {
+            inner: Some(TransferInner {
+                buffer,
+                channel,
+                target,
+            }),
+        }
+    }
+
+    /// Is this transfer complete?
+    pub fn is_complete(&self) -> bool {
+        let inner = self.inner.as_ref().unwrap();
+        inner.channel.event_occured(Event::TransferComplete)
+    }
+
+    /// Stop this transfer and return ownership over its parts
+    pub fn stop(mut self) -> (B, C, T) {
+        let mut inner = self.inner.take().unwrap();
+        inner.stop();
+
+        (inner.buffer, inner.channel, inner.target)
+    }
+
+    /// Block until this transfer is done and return ownership over its parts
+    pub fn wait(self) -> (B, C, T) {
+        while !self.is_complete() {}
+
+        self.stop()
+    }
+}
+
+impl<B, C: Channel, T> Drop for Transfer<B, C, T> {
+    fn drop(&mut self) {
+        if let Some(inner) = self.inner.as_mut() {
+            inner.stop();
+        }
+    }
+}
+
+/// This only exists so we can implement `Drop` for `Transfer`.
+struct TransferInner<B, C, T> {
+    buffer: B,
+    channel: C,
+    target: T,
+}
+
+impl<B, C: Channel, T> TransferInner<B, C, T> {
+    /// Stop this transfer
+    fn stop(&mut self) {
+        self.channel.disable();
+        atomic::compiler_fence(Ordering::Acquire);
```

Comment:
But the `self.channel.disable()` does a read, no? It is implemented as:
```rust
self.ch().cr.modify(|_, w| w.en().disabled());
```
And `modify` does a read-modify-write as far as I understand. I mean, it has to. How else would it set a single bit and leave the others unchanged?

Reply:
The purpose of this fence is to prevent operations on the buffer from getting moved to before the DMA is disabled, that happens in the write part. In `modify`, the read comes before the write, so the `Acquire` fence is, effectively, between the read and write. That way, it prevents the operations from being moved to before the read, but not the write.

Reply:
You are right of course.

What do you think of adding an `unsafe { ptr::read_volatile(&0); }` after the `channel.disable()`? That's what suggested in the Embedonomicon to force a memory read. That way we still have a single fence for DMA start and a single fence for DMA stop, which I think is less confusing than having multiple ones. Also we'd avoid the `SeqCst` fence, which sounds more expensive to me.

Reply:
To be honest, I don't think that `SeqCst` would have a noticeable difference (in performance) from `Acquire` here in specific. And I'm almost sure that pushing a variable to the stack, taking its address and reading it would probably be more expensive, so I would go with `SeqCst` which is even more readable than having an `unsafe { ptr::read_volatile(&0); }`.

Reply:
You're right, I'm getting lost in premature optimization. Still I think it would be preferable to just change this fence here to `SeqCst` ordering instead of introducing a second one and making it harder to be sure that we have a fence in every relevant path.

Reply:
Makes sense.